src/compas/geometry/transformations/quaternions.py

```diff
@@ -0,0 +1,324 @@
+
+    Returns
+    -------
+    list
+        A quaternion of length 1.0.
+    """
+    n = quaternion_norm(q)
+    if isclose(n, 0.0):
+        raise ValueError("The given quaternion has zero length.")
+    else:
+        return [x/n for x in q]
+
+
+def quaternion_is_unit(q):
+    """Checks if a quaternion is unit-length.
+    Parameters
+    ----------
+    q : list
+        Quaternion as a tuple of four real values [qw,qx,qy,qz].
+
+    Returns
+    -------
+    bool
+        True if the quaternion is unit-length, and False if otherwise.
+    """
+    n = quaternion_norm(q)
+    return isclose(n, 1.0)
+
+
+def quaternion_multiply(r, q):
+    """Multiplies two quaternions.
+    Parameters
+    ----------
+    r : list
+        Quaternion as a tuple of four real values [qw,qx,qy,qz].
+    q : list
+        Quaternion as a tuple of four real values [qw,qx,qy,qz].
+
+    Returns
+    -------
+    list
+        A quaternion p = r * q.
+
+    Note
+    ----
+    Multiplication of two quaternions r * q can be interpreted as applying rotation r to an orientation q,
+    provided both r and q are unit-length.
+    The result is also unit-length.
+    Multiplication of quaternions is not commutative!
+
+    .. _mathworld quaternion:
+    http://mathworld.wolfram.com/Quaternion.html
+    """
+    rw, rx, ry, rz = r
+    qw, qx, qy, qz = q
+    pw = rw*qw - rx*qx - ry*qy - rz*qz
+    px = rw*qx + rx*qw + ry*qz - rz*qy
+    py = rw*qy - rx*qz + ry*qw + rz*qx
+    pz = rw*qz + rx*qy - ry*qx + rz*qw
+    return [pw, px, py, pz]
+
+
+def quaternion_canonic(q):
+    """Converts a quaternion into a canonic form if needed.
+
+    Parameters
+    ----------
+    q : list
+        Quaternion as a tuple of four real values [qw,qx,qy,qz].
+
+    Returns
+    -------
+    list
+        A quaternion in a canonic form.
+
+    Note
+    ----
+    Canonic form means the scalar component is a non-negative number.
+
+    """
+    if q[0] < 0.0:
+        return [-x for x in q]
+    else:
+        return [x for x in q]
+
+
+def quaternion_conjugate(q):
+    """Conjugate of a quaternion.
+
+    Parameters
+    ----------
+    q : list
+        Quaternion as a tuple of four real values [qw,qx,qy,qz].
+
+    Returns
+    -------
+    list
+        A conjugate quaternion.
+
+    Note
+    ----
+    .. _mathworld quaternion conjugate
+    http://mathworld.wolfram.com/QuaternionConjugate.html
+
+    """
+    return [q[0], -q[1], -q[2], -q[3]]
+
+
+# ----------------------------------------------------------------------
+# quaternion_from_matrix(m) - already in compas/geometry/transformations/matrices.py
+# matrix_from_quaternion(q) - already in compas/geometry/transformations/matrices.py
+# ----------------------------------------------------------------------
+
+
+def quaternion_from_euler_angles(e, static=True, axes='xyz'):
+    """Returns a quaternion from Euler angles.
+
+    Parameters
+    ----------
+    euler_angles : list of float
```

Comment:
again just a question for discussion.
many of the inputs that just require something to be "iterable" are marked as `list`.
the `numpy` doc guidelines suggest to be as specific as possible, which in this case would be something like `iterable(3) of float` or perhaps just `float[3]`.

also, in the sense of "iterable" or `list` an `iterator` and a `list` are really not the same.
the "iterable" can be used only once in the code block because it gets consumed.
a `list` or `tuple` can be reused multiple times...

Reply:
@romanarust 
@gonzalocasas